### PR TITLE
Implement Transaction Isolation Controls for Concurrent Operations

### PR DESCRIPTION
**Pull request by kevinrutledge**
## Description
This PR adds transaction isolation controls to prevent race conditions in critical API operations. These changes ensure data consistency during concurrent access and modifications.

## Key Changes
- Added transaction isolation controls to route optimization endpoint
- Added transaction isolation controls to price comparison endpoint
- Added transaction isolation controls to shopping list operations
- Added comprehensive test scenarios for validation

## Technical Details
- Implemented REPEATABLE READ isolation level for operations requiring consistent views
- Wrapped critical database operations in proper transaction boundaries
- All changes maintain existing API contracts and responses

## Testing Done
Created test scenarios for validating concurrency behavior:
1. Route optimization with changing inventory
2. Price comparison during price updates
3. Shopping list creation with concurrent price changes

## Checklist
- [x] Code follows project style guidelines
- [x] Documentation has been updated
- [x] Test scenarios have been created
- [x] No breaking changes
- [x] Performance impact assessed

### CODE REVIEW

**Comment by TLMatthies** [comment]
Reviewed it, looks good. Great work.

**Review by TLMatthies — APPROVED** [review]
Looks good.

File: src/api/shopping.py
Comment (Mcray3000) [inline src/api/shopping.py:1]: TODO: fix practices to be consistent

**Comment by Mcray3000** [comment]
All of your changes look good Mr Kevin

**Pull request merged**